Create tutorial doc on FF.com for Collaborative Text Area using SharedString and React (#8081)
microsoft/FluidFramework: docs/content/docs/recipes/shared-string.md

---
title: Using SharedString with React
menuPosition: 4
author: scottn12
---

> In this tutorial, you'll learn how to use the `SharedString` distributed data structure (DDS) with [React](https://reactjs.org/). To learn more about `SharedString`, click [here](https://fluidframework.com/docs/data-structures/string/).

Reviewer: ```suggestion
In this tutorial, you'll learn how to use the `SharedString` distributed data structure (DDS) with [React](https://reactjs.org/) to create a collaborative text area. To learn more about `SharedString`, click [here](https://fluidframework.com/docs/data-structures/string/).
```
or something similar that adds what the sharedstring is used to create here

To jump ahead into the finished demo, check out the [SharedString example in our FluidExamples repo](TODO_URL_HERE).

> The following image shows a textarea open in four browser. The same text is in all four. The cursor is in the bottom textarea in one browser.

Reviewer: ```suggestion
The following image shows a textarea open in four browsers. The same text is in all four. The cursor is in the bottom textarea in one browser.
```

Reviewer: would "text area" instead of "textarea" be more appropriate here?  i would think the tutorial is more about a general text area rather than specifically a textarea element

    | `@fluidframework/tinylicious-client`   |Defines the connection to a Fluid service server and defines the starting schema for the [Fluid container][].|
    {.table}

    Run the following command to install the libraries.

    ```dotnetcli
    npm install @fluidframework/tinylicious-client fluid-framework
    ```

## Code the project

1. Open the file `\src\App.js` in your code editor. Delete all the default `import` statements except the one that imports `App.css`. Then delete all the markup from the `return` statement. The file should look like the following:

    ```js
    import "./App.css";

    function App() {
      return (

      );
    }

    export default App;
    ```

1. Add the following `import` statements. Note: `CollaborativeTextArea` and `SharedStringHelper` will be defined later.

    ```js
    import React from "react";
    import { TinyliciousClient } from "@fluidframework/tinylicious-client";
    import { SharedString } from "fluid-framework";
    import { CollaborativeTextArea } from "./CollaborativeTextArea";
    import { SharedStringHelper } from "./SharedStringHelper";
    ```

### Get Fluid Data

> 1. The Fluid runtime will bring changes made to the text from any client to the current client. But Fluid is agnostic about the UI framework. You can use a React hook to get the Fluid data, from the SharedString object, into the view layer (the React state). Add the following code below the `import` statements. This method is called when the application loads the first time, and the value that is returned form it is assigned to a React state property.

Reviewer: ```suggestion
1. The Fluid runtime will bring changes made to the text from any client to the current client, but Fluid is agnostic about the UI framework. You can use a React hook to get the Fluid data from the SharedString object into the view layer (the React state). Add the following code below the `import` statements. This method is called when the application loads the first time, and the returned value is assigned to a React state property.
```

    ```js
    let container;
    const containerId = window.location.hash.substring(1);
    if (!containerId) {
      container = (await client.createContainer(containerSchema)).container;
      const id = await container.attach();
      window.location.hash = id;
    } else {
      container = (await client.getContainer(containerId, containerSchema)).container;
    }
    ```

1. Replace `TODO 3` with the following code.

    ```js
    return container.initialObjects.sharedString;
    ```

1. Replace `TODO 4` with the following code. Note about this code:
    - By setting an empty dependency array at the end of the `useEffect`, it is ensured that this function only gets called once.
    - Since `setFluidSharedObjects` is a state-changing method, it will cause the React `App` component to immediately rerender.

    ```js
    React.useEffect(() => {
      getFluidData()
        .then(data => setSharedString(data));
    }, []);
    ```

1. Finally, replace `TODO 5` with the following code.

    ```js
    return sharedString;
    ```

### Move the Fluid Data to the view

> Inside the `App()` function, add the following code. Note about this code:

Reviewer: is it worth adding a TODO comment within the App method as a placeholder for this?

- The `sharedString` object returned from the code above is used to create a `SharedStringHelper` object, which is a class that provides simple APIs to interact with the `sharedString` object.

> - Next, the `SharedStringHelper` object is then passed into the `CollaborativeTextArea` React component, which integrates `SharedString` with the default `textarea` HTML element to allow co-editing through Fluid.

Reviewer: ```suggestion
- Next, the `SharedStringHelper` object is passed into the `CollaborativeTextArea` React component, which integrates `SharedString` with the default `textarea` HTML element to allow co-editing through Fluid.
```

```js
const sharedString = useSharedString();

if (sharedString) {
  return (
    <div className="app">
      <CollaborativeTextArea sharedStringHelper={new SharedStringHelper(sharedString)} />
    </div>
  );
} else {
  return <div />;
}
```

### Setup the SharedStringHelper Class

> As previously mentioned, the `SharedStringHelper` class is used to create simple APIs to make interactions with the `SharedString` object easier by handling the merge logic. To implement this class, follow the below instructions.

Reviewer: ```suggestion
As previously mentioned, the `SharedStringHelper` class provides simple APIs to make interactions with the `SharedString` object easier by handling the merge logic. To implement this class, follow the below instructions.
```

        newPosition = oldPosition + changeLength;
      }
      return newPosition;
    };
    ```

    Replace `TODO 3B` with the following code to handle a `REMOVE` operation.

    ```js
    transformPosition = (oldPosition) => {
      if (op.pos1 === undefined) {
        throw new Error("pos1 undefined");
      }
      if (op.pos2 === undefined) {
        throw new Error("pos2 undefined");
      }
      const changeStartPosition = op.pos1;
      const changeEndPosition = op.pos2;
      const changeLength = changeEndPosition - changeStartPosition;
      let newPosition;
      if (oldPosition <= changeStartPosition) {
        // Position is unmoved by the deletion if it is before the deletion's start
        newPosition = oldPosition;
      } else if (oldPosition > (changeEndPosition - 1)) {
        // Position is moved by the size of the deletion if it is after the deletion's end
        newPosition = oldPosition - changeLength;
      } else {
        // Position snaps to the left side of the deletion if it is inside the deletion.
        newPosition = changeStartPosition;
      }
      return newPosition;
    };

>     ```

Author: Felt it may be a good idea to break up TODO 3 into TODO 3A and 3B since otherwise it may be too much for a single code block

Reviewer: seeing this here makes me think this might be out of scope for this example and would be better suited as an import.  like, i don't think this is too different from delving into SharedString internals, which is more clearly out of scope.

Author: I think this makes sense. Only concern may be using an import from `fluid-experimental`. @tylerbutler , would you mind giving your opinion?

Reviewer 2: Yeah, experimental isn't great. That said, I don't want to block on that. Just go ahead and pull in the experimental import and file an issue to track making sure this function (or the whole package) is eventually moved out of experimental.

1. Add the following import statements and declare the `CollaborativeTextArea` component:

    ```js
    import React from "react";

    export const CollaborativeTextArea = (props) => {
      // TODO 1: Setup React state and references
      // TODO 2: Handle a change event in the textarea
      // TODO 3: Set the selection in textarea element (update the UI)
      // TODO 4: Store current selection from the textarea element in the React ref
      // TODO 5: Detect changes in sharedStringHelper and update React/UI as necessary
      // TODO 6: Create and configure a textarea element that will be used in App.js
    }
    ```

1. Replace `TODO 1` with the following code.

    ```js
    const sharedStringHelper = props.sharedStringHelper;  // Instance of SharedStringHelper class

    const textareaRef = React.useRef(null);  // Ref for HTML textarea element
    const selectionStartRef = React.useRef(0);  // Ref for start of selected text
    const selectionEndRef = React.useRef(0);  // Ref for end of selected text

    const [text, setText] = React.useState(sharedStringHelper.getText());  // State variable for text inside the textarea element
    ```

> 1. Replace `TODO 2` with the following code. Note about this code:

Reviewer: missing notes?

Author: Fixed in last commit 😅

    ```js
    return (
    <textarea
      rows={20}
        cols={50}
        ref={textareaRef}
        onBeforeInput={storeSelectionInReact}
        onKeyDown={storeSelectionInReact}
        onClick={storeSelectionInReact}
        onContextMenu={storeSelectionInReact}
        onChange={handleChange}
        value={text} />
    );
    ```

## Start the Fluid server and run the application

In the Command Prompt, run the following command to start the Fluid service. Note that `tinylicious` is the name of the Fluid service that runs on localhost.

```dotnetcli
npx tinylicious
```

Open a new Command Prompt and navigate to the root of the project; for example, `C:/My Fluid Projects/shared-string-tutorial`. Start the application server with the following command. The application opens in your browser. This may take a few minutes.

```dotnetcli
npm run start
```

Paste the URL of the application into the address bar of another tab or even another browser to have more than one client open at a time. Edit the text on any client and see the text change and synchronize on all the clients.

## Next steps

> - Try extending the demo with more Fluid DDSes and a more complex UI.

Reviewer: tips 1 and 3 here are kind of the same
